### ISSUE TITLE
parsing issue affecting content assist

### ISSUE THREAD

**Issue by msmihai** [eclipse/xtext-core#1132]
Given the following grammar definition

`grammar org.xtext.example.mydsl.MyDsl with org.eclipse.xtext.common.Terminals hidden(SL_COMMENT, WS)

generate myDsl "http://www.xtext.org/example/mydsl/MyDsl"

World:
    {World}
    fooBars+=FooBar*
    ;

FooBar:
    Foo|Bar
;

Foo:
    "Foo" name=ID '{'
      fooDef=FooDef
    '}'
;

FooDef:
    "FooDef" name=ID '{'
    "barRef" '=' barRef=[Bar|ID] ';'
    '}';

Bar:
    "Bar" name=ID '{'
    '}';`

The following code works just fine:

`Bar bar1{}

Foo foo1
{
	FooDef fd
	{
		barRef=bar1;
	}
}`

Reversing the definitions of bar1 and foo1 only seems like working fine (it validates):

`Foo foo1
{
	FooDef fd
	{
		barRef=bar1;
	}
}

Bar bar1{}`

but actually, if you try to use the content assist after "bar Ref=" you will see it doesn't work.

The internal parser (antlr?) is confused because of the closing curly bracket of foo1 so everything after the foo1 definition is not parsed anymore.

The workaround is:

`FooDef:
    "FooDef" name=ID '{'
    	x=X
    '}';
    
X:
	"barRef" '=' barRef=[Bar|ID] ';'
;   `

**Comment by sterpumihai** [comment]
> I'll likely work better if you extract a rule or rule fragment for the `FooDef` body (containing `"barRef" '=' barRef=[Bar|ID] ';'`) Can you give that a shot?

Hi,

I don't understand, isn't this the same workaround I provided in my initial post?

**Comment by szarnekow** [comment]
Yes, it is. Sorry for the noise. The only different is the hint to use a fragment which would be transparent for the produced AST.
